### ISSUE TITLE
Meteor, cordova, DDP, accounts package, autoupdate package, and remote servers. Cannot make Meteor/Cordova use a remote server for DDP.

### ISSUE THREAD

**Issue by darrynten** [meteor/meteor#5469]
Greetings,

I've been trying to solve the following problem for a few days now, and it's been driving me absolutely crazy.

I have a (1.2) meteor application, deployed at `http://some.application.com:3000`. It works great, and does what it is supposed to do. The application uses several packages, the ones that I think are related to this issue are `autoupdate` and the `accounts` package (which loads its own bunch of stuff).

Our directive is to turn this webapp into an android app, something we've been told meteor can do "quite easily". On the surface it seems like it's a simple case of `meteor run android-device --mobile-server http://some.application.com:3000 --settings settings.json --verbose`, however this doesn't do what I expect it to do.

Meteor decides to do the DDP connection on `10.0.2.2` (for whatever reason), and no matter what env variables I set I end up in the same situation.

It's important to note that the application has not been written using the `DDP.connect(url)` method anywhere [[docs](http://docs.meteor.com/#/full/ddp_connect)], so everything relies on the primary DDP connection (which I suspect might be one of the bigger causes of our problem).

For the record, here is my startup script. I got pretty desperate and added many, many env vars, and haven't had any luck for any combination thereof.

```bash
#!/bin/bash
export AWS_REGION=xxx
export AWS_BUCKET=xxx
export MONGO_URL=mongodb://some.application.com:27017/application
export QUEUE_ADDRESS=http://some.application.com
export AWS_ACCESS_KEY_ID=xxx
export AWS_SECRET_ACCESS_KEY=xxx
export ROOT_URL=http://some.application.com:3000
export DDP_DEFAULT_CONNECTION_URL=http://some.application.com:3000
export MOBILE_DDP_URL=http://some.application.com:3000 
export MOBILE_ROOT_URL=http://some.application.com:3000

# Let's go
meteor run android-device --mobile-server http://some.application.com:3000 --settings settings.json --verbose
```

Running it locally, on mobile or desktop, (via `localhost:3000` with port forwarding, or any other internal IP (`10.x.x.x`, `192.x.x.x`) works absolutely fine. It even works with the remote AWS, Queue and DB.

According to all the documentation the `--mobile-server` switch should sort things out. It doesn't. I've tried it with and without an `=`, wrapped in quotes, all possible ways of defining it.

Looking at the `<head>` of my document I see the following code getting injected

```js
__meteor_runtime_config__ = JSON.parse(decodeURIComponent("%7B%22meteorRelease%22%3A%22METEOR%401.2.0.2%22%2C%22PUBLIC_SETTINGS%22%3A%7B%22verifiedLogin%22%3Afalse%2C%22enableFacebookAuth%22%3Afalse%2C%22enableTwitterAuth%22%3Afalse%2C%22enableGoogleAuth%22%3Afalse%2C%22cdnUrlWithTrailingSlash%22%3A%22http%3A%2F%2Fdev.cdn.some.application.com%2F%22%2C%22ga%22%3A%7B%22id%22%3A%22UA-XXXXXX-1%22%7D%7D%2C%22ROOT_URL%22%3A%22http%3A%2F%2Flocalhost%3A3000%22%2C%22ROOT_URL_PATH_PREFIX%22%3A%22%22%2C%22appId%22%3A%228emj6c37j3fdoz5qmp%22%2C%22accountsConfigCalled%22%3Atrue%2C%22autoupdateVersion%22%3A%222b3acf7aa3ddef802ddf661d3b3986319aad5122%22%2C%22autoupdateVersionRefreshable%22%3A%22b00197cdb5345434d03d9a2503906349ff7854e2%22%2C%22autoupdateVersionCordova%22%3A%223644168d46bc4597d0b2d8c39e366890f6725f52%22%2C%22DDP_DEFAULT_CONNECTION_URL%22%3A%22http%3A%2F%2Flocalhost%3A3000%22%7D"));

if (/Android/i.test(navigator.userAgent)) {
  // When Android app is emulated, it cannot connect to localhost,
  // instead it should connect to 10.0.2.2
  // (unless we're using an http proxy; then it works!)
  if (!__meteor_runtime_config__.httpProxyPort) {
    __meteor_runtime_config__.ROOT_URL = (__meteor_runtime_config__.ROOT_URL || '').replace(/localhost/i, '10.0.2.2');
    __meteor_runtime_config__.DDP_DEFAULT_CONNECTION_URL = (__meteor_runtime_config__.DDP_DEFAULT_CONNECTION_URL || '').replace(/localhost/i, '10.0.2.2');
  }
}
```

The UrlDecoded version of that string is as follows

```json
{
    "meteorRelease": "METEOR@1.2.0.2",
    "PUBLIC_SETTINGS": {
        "verifiedLogin": false,
        "enableFacebookAuth": false,
        "enableTwitterAuth": false,
        "enableGoogleAuth": false,
        "cdnUrlWithTrailingSlash": "http://dev.cdn.application.com/",
        "ga": {
            "id": "UA-XXXXXX-1"
        }
    },
    "ROOT_URL": "http://localhost:3000",
    "ROOT_URL_PATH_PREFIX": "",
    "appId": "jfdjdjdjdjdjjdjdjdjjd",
    "accountsConfigCalled": true,
    "autoupdateVersion": "2b3acf7aa3ddef802ddf661d3b3986319aad5122",
    "autoupdateVersionRefreshable": "b00197cdb5345434d03d9a2503906349ff7854e2",
    "autoupdateVersionCordova": "3644168d46bc4597d0b2d8c39e366890f6725f52",
    "DDP_DEFAULT_CONNECTION_URL": "http://localhost:3000"
}
```

This is strange because I have no entries of localhost anywhere. 

Booting the app tells me: `App running at: http://site.some.application.com`, but no connections are made in the network inspector.

Grepping through the code shows me that the only places where `__meteor_runtime_config__` is mentioned is in the `autoupdate` package.

Further investigation lead me to [this issue](https://github.com/meteor/meteor/issues/3815) #3815 which linked to [this fix](https://github.com/rclai/meteor-app-with-remote-server), but after I implemented it (the changes to the autoupdate package) I was still faced with the same problem (although hot code fixes stopped coming through from my local machine)

Even more investigation lead me to believe that the remote DDP server could be changed [like this](https://github.com/gwendall/meteor-remote-ddp), but unfortunately this solution doesn't work with Cordova.

I tried settings `HTTP_PROXY` as the comment "unless we're behind a proxy" in the `<head>` script lead me to believe this might be a quick fix, but I didn't have any luck with this.

I tried removing the accounts package, but have not had any luck in this regard.

**Main Question** Is there any suggested way to allow a Cordova wrapped Meteor application to connect to an arbitrary server, and allow a DDP connection to same? 

The accounts package is (most likely) needed. I suppose auto-updates aren't thaaat crucial, although they do help in terms of not having to regularly release code to the various app stores.

Things I've tried:

* Removing accounts package
* Remove autoupdates
* Modifying autoupdates to point to remote DDP
* Using the `remote-ddp` package
* Forcing `__meteor_runtime_config__` overrides
* Setting a proxy
* Environment variables
* And several other thousand things

Related issues (Going back to Jan 2015) are:

* [How can DDP_DEFAULT_CONNECTION_URL be set?](https://github.com/meteor/meteor/issues/3852) #3852 - Shows difficulty in connecting to remote meteor servers, and touches on how the `autoupdate` package affects things.
* [Dont' start local server when using option --mobile-server ](https://github.com/meteor/meteor/issues/3727) #3727 - This shows a case of the `--mobile-server` becoming `10.0.2.2`
* [Meteor mobile build is not changing DDP_DEFAULT_CONNECTION_URL](https://github.com/meteor/meteor/issues/4396) #4396 - This shows an apparent fix, but this doesn't work for me at all
* [Ability to pass an alternative DDP connection to autoupdate](https://github.com/meteor/meteor/issues/3815) #3815 - This shows the confusion that comes from the official documentation, and lead me to the `autoupdate` package "fix" that I linked earlier
* [MOBILE_ROOT_URL and MOBILE_DDP_URL are ignored on meteor build](https://github.com/meteor/meteor/issues/4581) #4581 - This shows how meteor build ignores these env vars
* [Can't build mobile app with different DDP server](https://github.com/meteor/meteor/issues/4412) #4412 - This shows others having difficulty with the same problem, with the response asking for PRs around the issue
* [Meteor Accounts only authenticates DDP, not HTTP](https://github.com/meteor/meteor/issues/3390) #3390 - This shows that auth via `meteor-accounts` can only happen via DDP, and not HTTP
* [Built apps cannot connect to the given --server: they keep failing to connect](https://github.com/meteor/meteor/issues/3698) #3698 - This shows other users having the same issue on iOS, although they do report having success with connecting to a local server, which I also have success with, but there is no mention of success with a remote server. The fix appears to be deploying through meteor to `some-app.meteor.com` but this isn't an option for us.

Contents of `.meteor/packages`

```bash
aldeed:autoform@=4.2.2
aldeed:collection2@2.5.0
aldeed:simple-schema@1.3.3
aldeed:tabular@1.4.1
autoupdate@1.2.3
biasport:facebook-sdk@0.2.2
blaze@2.1.3
check@1.0.6
edgee:slingshot@0.7.1
iron:router@1.0.12
jquery@1.11.4
juliancwirko:s-alert@3.1.1
juliancwirko:s-alert-slide@3.1.0
lookback:seo@1.1.0
matteodem:easy-search@1.6.4
meteor@1.1.9
meteorhacks:fast-render@2.10.0
meteorhacks:subs-manager@1.6.2
mobile-experience@1.0.1
momentjs:moment@2.10.6
mquandalle:jade@0.4.4
multiply:iron-router-progress@1.0.2
---
internal packages (one of which includes accounts)
---
reactive-dict@1.1.2
reactive-var@1.0.6
reywood:iron-router-ga@0.7.1
session@1.1.1
standard-minifiers@1.0.1
templating@1.1.4
tracker@1.0.9
underscore@1.0.4
underscorestring:underscore.string@3.2.2
utilities:avatar@0.9.2
```

I can provide the contents of my `versions` file if you feel that will help.

**TL;DR - Is there any suggested way to allow a Cordova wrapped Meteor application to connect to an arbitrary server, and allow a DDP connection to same?**

Any help or pointers around this issue would be much appreciated. Please let me know if there is any other information you may need to assist in this regard.

Many thanks

**Comment by martijnwalraven** [comment]
`meteor run android-device --mobile-server http://some.application.com:3000` should indeed connect to a server running on some.application.com. The issue here is likely that although this connection initially succeeds, autoupdate then kicks in and reloads the app from some.application.com. This replaces index.html with a version where ROOT_URL has been configured as localhost:3000 (the default).

The solution is to also configure the server with the right connection address. This happens automatically if you're using `meteor deploy`, but when deploying yourself you'll have to make sure to define `ROOT_URL`.

(I don't see where this is used in your case, but 10.0.2.2 is a special address the Android emulator uses to connect to the host machine, see [documentation here](http://developer.android.com/tools/devices/emulator.html#networkaddresses)).

**Comment by slavik0329** [comment]
I'm having this same exact issue. No matter what env variable I set and how I set them, I still get the device trying to connect to DDP on port 10.0.2.2

**Comment by slavik0329** [comment]
Okay, I found the issue I was having. Apparently if you do not uninstall the android application from the phone first, it won't overwrite the DDP URL settings. Uninstalling and rerunning meteor fixes this issue.

**Comment by lukasvan3l** [comment]
Just had this same problem, uninstalling and rerunning indeed fixes the problem, but this took me way too much time to get there. Can this be solved in Meteor so that each `meteor run x-device` refreshes the DDP URL setting?

**Comment by martijnwalraven** [comment]
I only found out about this recently, but I suspect https://github.com/meteor/meteor/issues/5639 is the issue here.

**Comment by romaroma** [comment]
It seems that this directive is not working as expected:
`--mobile-server http://example.com:3000`

we recently noticed that our mobile app tries to connect to following address:
`http://example.com`
insted of
`http://example.com:3000`

**Comment by martijnwalraven** [comment]
I believe this has been fixed in Meteor 1.3! A new beta is now available, please give it a try: https://github.com/meteor/meteor/issues/6266.

**Comment by rajpa** [comment]
I am facing this dreaded 10.0.2.2 problem again.. How do I fix this? Please help.

https://forums.meteor.com/t/signed-release-version-apk-not-working/33802/13

**Comment by abernix** [comment]
@rajpa It looks like I answered your question in the other issue you opened (#8333) and you confirmed it was fixed by Meteor 1.4.2.6.

    meteor update --release 1.4.2.6

Glad it works for you now!
